datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
+            #   'creationDate': '2022-07-20'}
+
+            # sort by version, so we can rely on first entry to refer to latest
+            # Note, that ('versionNumber', 'versionMinorNumber', 'versionState')
+            # would look like this:
+            # (None, None, 'DRAFT'), (2, 0, 'RELEASED'), (1, 0, 'RELEASED')
+            # and we need a possible DRAFT to have the biggest key WRT sorting.
+            self._dataset.sort(key=lambda v:
+                                   (v.get('versionNumber') or sys.maxsize,
+                                    v.get('versionMinorNumber') or sys.maxsize),
+                               reverse=True)
+
+        return self._dataset
+
+    @property
+    def files_all(self):
+        """Cache of files available from dataverse.
+
+        For quick lookup and deduplication, this is a dict {id: remote_path}
+        where remote_path is a `Path` build from the file's 'directoryLabel'
+        (if any) and 'filename'.
+        Currently, this is all we need, hence don't bother carrying and updating
+        all other metadata.
+        """
+
+        if self._files_all is None:
+            self._files_all = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f_l in [version['files'] for version in self.dataset]
+                for f in f_l}
+
+        return self._files_all
+
+    @property
+    def files_latest(self):
+        """Cache of files in the latest version of the dataverse dataset.
+
+        This refers to the DRAFT version (if there is any) or the latest
+        published version otherwise. That's the version pushes go into. Hence,
+        this is needed to determine whether we need to replace a file on
+        upload, while the complete list in `self.files_all` is relevant for key
+        retrieval (which can come from earlier dataverse versions of the
+        dataset).
+        """
+
+        if self._files_latest is None:
+            # Latest version in self.dataset is first entry.
+            self._files_latest = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f in self.dataset[0]['files']}
+
+        return self._files_latest
+
+    def remove_from_filelist(self, id):
+        """Update self.filelist after removal"""
+        if self._files_all is None:
+            self.message(f"FAILED Removing {id} from file list failed. "
+                         f"'_filelist' does not exist.",
+                         type='debug')
+            return  # not yet a reason to crash
+        self._files_all.pop(id)
+        try:
+            self._files_latest.pop(id)
+        except KeyError:
+            # This is alright, we removed from _all, it may not have been in
+            # latest, just calling here to be sure we are consistent between the
+            # two.
+            pass
```

Comment:
Reviewing linearily from the top of the code file, it is still unclear to me why there are `files_latest` *and* `files_all`.